zebra-network/src/peer/client.rs

```diff
@@ -438,7 +439,7 @@ impl Client {
                 // Heartbeat task is still running.
                 Ok(())
             }
-            Poll::Ready(Ok(())) => {
+            Poll::Ready(Ok(_)) => {
```

Comment:
This and the panic below it generates me some doubts. Do you think we should return `Ok` here and `PeerError` below to avoid the panic ?

Reply:
Since we added a result to the JoinHandle, the type is now `Poll<Result<Result<(), PeerError>, JoinError>`.

There's a few different cases we need to handle:
- if the JoinError is_cancelled(), then some other task told the heartbeat task to exit, so we can return PeerError::HeartbeatTaskCancelled (you can make a new variant if you need to)
- if the JoinError is_panic(), then we need to re-panic here, we can't recover from a panic. There's a good example of how to do that in the JoinError try_into_panic() docs
- if the inner result is a PeerError, we can return that error
- if not, we can return peererror::HeartbeatTaskExited
- otherwise the code stays the same

We missed the cancelled case when we originally wrote this code, it might be a new tokio feature, I'm not sure.

Reply:
this is just an idea, i am ok on doing it  different https://github.com/ZcashFoundation/zebra/pull/4756/commits/90526e31fe78fbdd3dd43eab2cc4a5315fc36f89

Basically added an argument to `HeartbeatTaskExited`. Let me know how do you see it.